docs/design/moderation.md

```diff
@@ -0,0 +1,138 @@
+
+### Service Availability Considerations:
+- The proposed design would use a "fail-closed" approach for safety - if the moderation service encounters an error, content would be blocked by default.
+- This ensures potentially problematic content doesn't slip through due to technical issues.
+- However, this means if moderation APIs are unavailable, all LLM interactions would be disabled by default.
+
+**Open Questions:**
+- Should there be an option for fail-open in certain trusted environments?
+- How can we implement graceful degradation instead of complete service shutdown?
+
+## Proposed Moderation Flow
+
+The proposed implementation would include three primary moderation points:
+
+1. **User Input Moderation**:
+   - When a user sends a message to an AI bot, the message would be checked by the moderation service.
+   - If flagged, the user would receive a notification and the message would not be processed by the LLM.
+   - This happens in `processUserRequestToBot()` in `conversations.go`.
+
+2. **LLM Response Moderation**:
+   - When an LLM generates a response, the complete response would be checked before being displayed to users.
+   - If flagged, users would see a generic message instead of the problematic content.
+   - This happens in `streamResultToPost()` in `post_processing.go`.
+
+3. **Image Moderation**:
+   - Images attached to messages would be moderated before being sent to vision-capable LLMs.
+   - Images that fail moderation would not be included in the context sent to the LLM.
+   - This happens in `PostToAIPost()` in `post_processing.go`.
+
+### Review Mechanism Considerations:
+- If content exceeds the configured threshold, it would be automatically rejected without any human review process.
+- This binary approach (allow/reject) may be too rigid for some use cases.
+
+**Open Questions:**
+- Should we implement a review queue for borderline content?
+- Should moderators have the ability to override automated decisions?
+- How would a review process be integrated with Mattermost's permission system?
+
+**Potential Enhancements:**
+- **Human-in-the-Loop Review**: We could implement a review queue where content that falls within a "gray zone" (e.g., between 0.6-0.8 on the threshold) would be sent for human review before final determination.
+- **Graduated Response**: Rather than a binary allow/block decision, we could implement multiple response levels:
+  - Allow (below threshold)
+  - Warn user but allow (borderline)
+  - Block but allow override with explanation (moderately above threshold)
+  - Block completely (significantly above threshold)
+- **Feedback Loop**: We could create a mechanism for users to contest moderation decisions, providing a learning loop for the system and reducing false positives over time.
+
+### User Feedback Considerations:
+- Users receive generic messages when their content is flagged, without specifics about why.
+
+**Open Questions:**
```

Comment:
Let's focus on a minimum viable product for the first iteration of this.  We can prioritize enhancements after we get customer feedback.